gpflow/utilities/utilities.py

```diff
@@ -246,22 +247,123 @@ def clear_bijector(bijector, _, state):
+        value = module.a.b.c[0].d[1000]
+        c0, dname, dindex = _get_last_attr_spec(module, "a.b.c[0].d[1000]")
+        assert c0 is module.a.b.c[0]
+        assert dname == "d"
+        assert dindex == "1000"
+
+    :param parent: A python object with a nested structure.
+    :param attr_path: An attribute path.
 
+    :returns: The value stored in the nested object by the attribute path.
+    """
 
-def deepcopy_components(input_module: M) -> M:
+    # Regexp extracts attribute names and indices if available.
+    # Outer brackets represent repeated pattern inside and contains 3 groups:
+    # - '(\w+)' is a group for attribute name.
+    # - '(\[(-?[\s\d]+)\])' is the index group and has a subgroup '(-?[\s\d]+)', it may or may
+    #   not appear in the pattern
+    # - '(-?[\s\d]+)' is an index without squared brackets
+    # - Last part of the expression reflects that the period delimeter cannot be last in a path.
+    restr = r"((\w+)(\[(-?[\s\d]+)\])?(?<!\.))+"
+    curr_parent = parent
+    attrs = re.findall(restr, attr_path)
+    for _, attr_str, _, index_str in attrs[:-1]:
+        curr_parent = _get_by_name_index(curr_parent, attr_str, index_str)
+    _, attr_str, _, index_str = attrs[-1]
+    return curr_parent, attr_str, index_str
+
+
+def getattr_by_path(target: object, attr_path: str) -> Any:
+    """Get a value of nested attribute by a string path.
+
+    :param target: Root object that contains nested attribute
+    :param attr_path: String type value that represents path to an attribute.
+
+    Example:
+        k = gpflow.kernels.Matern52()
+        m = gpflow.models.GPR(..., kernel=kernel)
+        lengthscale = getattr_by_path(m, "kernel.lengthscale")
+    """
+    try:
+        descendant, attr, index = _get_last_attr_spec(target, attr_path)
+        return _get_by_name_index(descendant, attr, index)
+    except ValueError as error:
+        raise ValueError(f"Cannot get value at path '{attr_path}'\n"
+                         f"Got error: {error}")
+
+
+def setattr_by_path(target: object, attr_path: str, value: Any):
+    """Set `value` by a given path to a nested attribute.
+
+    :param target: Root object that contains a nested attribute.
+    :param attr_path: String type value that represents path to the attribute.
+    :param value: Value to assign to the attribute.
+
+    Example:
+        k = gpflow.kernels.Matern52()
+        m = gpflow.models.GPR(..., kernel=kernel)
+        setattr_by_path(m, "kernel.lengthscale", tf.constant(1.0, dtype=...))
+    """
+    try:
+        descendant, attr, index = _get_last_attr_spec(target, attr_path)
+        return _set_by_name_index(descendant, value, attr, index)
+    except (AttributeError, IndexError, ValueError) as error:
+        raise ValueError(f"Cannot assign value at path '{attr_path}'\n"
+                         f"Got error: {error}")
+
+
+M = TypeVar("M", bound=tf.Module)
+
+
+def deepcopy(input_module: M, freeze: bool = False) -> M:
```

Comment:
`copy.deepcopy` [already has an optional argument](https://docs.python.org/3.8/library/copy.html). I suggest making a separate function `freeze` which deepcopy-s and freezes in one go